src/arcsf/config/experiment.py

```diff
@@ -0,0 +1,304 @@
+import os
+import warnings
+from copy import copy
+from itertools import product
+
+import wandb
+import yaml
+from jinja2 import Environment, FileSystemLoader
+
+import arcsf.constants
+from arcsf.config.config import Config
+from arcsf.data.config import DataConfig
+from arcsf.models.config import ModelConfig
+
+
+def _listify(obj: object):
+    """Takes an object. If a list, returns it. If not, returns a list containing
+    the object.
+
+    Args:
+        obj: Object to ensure is a list.
+
+    Returns:
+        Either the original object if a list or a list containing the object.
+    """
+    if isinstance(obj, list):
+        return obj
+    return [obj]
+
+
+def _make_config_name(
+    top_config_name: str,
+    n: int,
+) -> str:
+    return f"{top_config_name}/experiment_{n}"
+
+
+def _generate_combos_from_dict(
+    combo_dict: dict,
+    full: bool,
+) -> dict:
+    # Convert combo dict into dict of combos
+    sweep_dict_keys, sweep_dict_vals = zip(*combo_dict.items())
+    combinations = [
+        dict(zip(sweep_dict_keys, v)) for v in product(*list(sweep_dict_vals))
+    ]
+
+    # Initialise train_type
+    train_type = "retain"
+    if full:
+        train_type = "full"
+
+    # Add train_type to each combination
+    for n, _ in enumerate(combinations):
+        combinations[n]["train_type"] = train_type
+
+    # Return
+    return combinations
+
+
+def generate_combos_from_dict(
+    combo_dict: dict,
+    full: bool,
+    wandb_kwargs: dict,
+) -> dict:
+    # Generate combinations
+    combinations = _generate_combos_from_dict(combo_dict, full)
+
+    # Process model kwargs
+    for n, _ in enumerate(combinations):
+        combinations[n]["hyperparameter_config"] = combinations[n]["model_config"][1]
+        combinations[n]["model_config"] = combinations[n]["model_config"][0]
+        combinations[n] = {**combinations[n], **wandb_kwargs}
+
+    # Return
+    return combinations
+
+
+def generate_experiment_configs(
+    top_config_name: str,
+) -> None:
+
+    # Read in yaml file
+    with open(
+        os.path.join(arcsf.constants.EXPERIMENT_CONFIG_DIR, f"{top_config_name}.yaml")
+    ) as f:
+        top_config = yaml.safe_load(f)
+
+    # Loop over, construct combo dict
+    combo_dict = {}
+    for key, value in top_config["combinations"].items():
+        combo_dict[key] = _listify(value)
+
+    # Construct same dict with full dataset
+    combo_dict_full = copy(combo_dict)
```

Comment:
`deepcopy` would be safer here (just `from copy import deepcopy`)

Reply:
Have replaced `copy` with `deepcopy`